src/bootstrap/replication/peer_discovery.rs

```diff
@@ -0,0 +1,160 @@
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn gossip_addr_to_http_addr(&self, gossip_addr: &str) -> Option<String> {
+        let parts: Vec<&str> = gossip_addr.split('/').collect();
+        if parts.len() >= 3 && parts[1] == "ip4" {
+            let ip = parts[2];
+            // When we have just the IP address, we'll assume HTTP and the default port
+            Some(format!("http://{}:{}", ip, DEFAULT_RPC_PORT))
+        } else {
+            None
+        }
+    }
+
+    fn handle_contact_info(&self, contact_info: ContactInfoBody) {
+        let peer_id = match PeerId::from_bytes(&contact_info.peer_id) {
+            Ok(p) => p,
+            Err(_) => {
+                warn!("Invalid peer id in contact info");
+                return;
+            }
+        };
+
+        let http_addr = if !contact_info.announce_rpc_address.is_empty() {
+            // This peer is announcing an RPC address, use that one
+            contact_info.announce_rpc_address
+        } else if !contact_info.gossip_address.is_empty() {
+            // Fallback to deriving from gossip address
+            match self.gossip_addr_to_http_addr(&contact_info.gossip_address) {
+                Some(addr) => addr,
+                None => {
+                    warn!(
+                        contact_info = %serde_json::to_string(&contact_info).unwrap_or_else(|_| "failed to serialize".to_string()),
+                        "Invalid gossip address format, not adding peer",
+                    );
+                    return;
+                }
+            }
+        } else {
+            warn!(
+                contact_info = %serde_json::to_string(&contact_info).unwrap_or_else(|_| "failed to serialize".to_string()),
+                "No RPC or gossip address provided in contact info, not adding peer"
+            );
+            return;
+        };
+
+        // Cooldown check
+        if let Some(ts) = self.incompatible_peers.lock().unwrap().get(&peer_id) {
+            if ts.elapsed() < INCOMPATIBLE_PEER_COOLDOWN {
+                return;
+            }
+        }
+
+        // Spawn validation (metadata height match) via RpcClientsManager::add_new_peer
+        let rpc_manager = self.rpc_manager.clone();
+        let incompatible = self.incompatible_peers.clone();
+        info!(peer = %peer_id, addr = %http_addr, "Discovered potential replication peer");
+        tokio::spawn(async move {
+            match rpc_manager.add_new_peer(http_addr.clone()).await {
+                Ok(true) => {
+                    info!(peer = %peer_id, addr = %http_addr, "Added new replication peer")
+                }
+                _ => {
+                    warn!(peer = %peer_id, addr = %http_addr, "Peer incompatible or validation failed; cooling down");
+                    incompatible.lock().unwrap().insert(peer_id, Instant::now());
```

Comment:
I wonder if it's worth separating unreachable vs doesn't have snapshot. But in practice, I guess it doesn't matter much why it failed?